pkg/alertmanager/multitenant.go

```diff
@@ -250,29 +315,150 @@ func createMultitenantAlertmanager(cfg *MultitenantAlertmanagerConfig, fallbackC
+			Help: "Current number of configurations owned by the alertmanager",
+		}),
+	}
+
+	if cfg.ShardingEnabled {
+		if err := enableSharding(am, ringStore); err != nil {
+			return nil, errors.Wrap(err, "setup alertmanager sharding ring")
+		}
 	}
 
+	// Initialize the top-level metrics.
+	am.SyncTotal.WithLabelValues(reasonInitial)
+	am.SyncTotal.WithLabelValues(reasonPeriodic)
+	am.SyncTotal.WithLabelValues(reasonRingChange)
+
 	if registerer != nil {
 		registerer.MustRegister(am.alertmanagerMetrics)
 	}
 
-	am.Service = services.NewTimerService(am.cfg.PollInterval, am.starting, am.iteration, am.stopping)
-	return am
+	am.Service = services.NewBasicService(am.starting, am.run, am.stopping)
+
+	return am, nil
 }
 
-func (am *MultitenantAlertmanager) starting(ctx context.Context) error {
-	// Load initial set of all configurations before polling for new ones.
-	am.syncConfigs(am.loadAllConfigs())
+func (am *MultitenantAlertmanager) starting(ctx context.Context) (err error) {
+	defer func() {
+		if err == nil || am.subservices == nil {
+			return
+		}
+
+		if stopErr := services.StopManagerAndAwaitStopped(context.Background(), am.subservices); stopErr != nil {
+			level.Error(am.logger).Log("msg", "failed to gracefully stop alertmanager dependencies", "err", stopErr)
+		}
+	}()
+
+	if am.cfg.ShardingEnabled {
+		if am.subservices, err = services.NewManager(am.ringLifecycler, am.ring); err != nil {
+			return errors.Wrap(err, "failed to start alertmanager's subservices")
+		}
+
+		if err = services.StartManagerAndAwaitHealthy(ctx, am.subservices); err != nil {
+			return errors.Wrap(err, "failed to start alertmanager's subservices")
+		}
+
+		am.subservicesWatcher = services.NewFailureWatcher()
+		am.subservicesWatcher.WatchManager(am.subservices)
+
+		// We wait until the ring is the JOINING state, once it does we know that tokens are assigned to this instance and we'll be ready to perform an initial sync of configs.
+		level.Info(am.logger).Log("waiting until alertmanager is JOINING in the ring")
+		if err = ring.WaitInstanceState(ctx, am.ring, am.ringLifecycler.GetInstanceID(), ring.JOINING); err != nil {
+			return err
+		}
+		level.Info(am.logger).Log("msg", "alertmanager is JOINING in the ring")
+	}
+
+	// At this point, if sharding is enabled, the instance is registered with some tokens
+	// and we can run the initial iteration to sync configs. If no sharding is enabled we load _all_ the configs.
+	if err := am.iteration(ctx, reasonInitial); err != nil {
+		return err
+	}
+
+	if am.cfg.ShardingEnabled {
+		// With the initial iteration now completed, we should have loaded all assigned alertmanager configurations to this instance. We can switch it to ACTIVE and start serving requests.
+		if err := am.ringLifecycler.ChangeState(ctx, ring.ACTIVE); err != nil {
+			return errors.Wrapf(err, "switch instance to %s in the ring", ring.ACTIVE)
+		}
+
+		// Wait until the ring client detected this instance in the ACTIVE state.
```

Comment:
Given the above operation to change state executes async, there could be a chance where the ring client does not take the instance as active yet. With this, we can guarantee it.

Reply:
Not sure I still understand why we need it. Let's say we remove this piece of code. The `run()` function will run and the `loadAndSyncConfigs()` will be re-triggered as soon as the ring client detects the switch to `ACTIVE` (because `HasReplicationSetChanged()` will detect it has a change).

Reply:
> will be re-triggered as soon as the ring client detects the switch to ACTIVE (because HasReplicationSetChanged() will detect it has a change

It's exactly that. We're trading off that extra sync which should not be needed for waiting a little bit more for the instance to be active. 

I think that additionally If we experience slow reads for the ring state this can become a problem where we know this instance is active and is part of the replication set but is not being detected. By waiting, we're guaranteeing we were able to read back its state before moving it on. 

I originally saw it as part of the storegateway and I thought it seemed like a sensible thing to include as well - I like the idea of guaranteeing that we won't need that initial sync because we waiting to read it was active.